### PR TITLE
define PerformanceObserver#takeRecords()

### LINKED ISSUE #74: Add support for flushing observer via `takeRecords()`
Related discussion in https://github.com/w3c/performance-timeline/issues/66.

**Use case:** as a developer I want to flush and beacon all pending records when page visibility changes (e.g. transitions to 'hidden') or is about to be unloaded. Today, this is not possible.

To stay consistent with other `<*>Observer` patterns, we should support `takeRecords()`:
- [MutationObserver#takeRecords](https://dom.spec.whatwg.org/#dom-mutationobserver-takerecords)
- [IntersectionObserver#takeRecords](https://wicg.github.io/IntersectionObserver/#dom-intersectionobserver-takerecords)

### PR DESCRIPTION
**Pull request by igrigorik**
Closes #74.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/performance-timeline/pull/98.html" title="Last updated on Dec 8, 2017, 10:37 PM GMT (9154b82)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/performance-timeline/98/43cf7a3...9154b82.html" title="Last updated on Dec 8, 2017, 10:37 PM GMT (9154b82)">Diff</a>

### CODE REVIEW

**Comment by igrigorik** [comment]
@npm1 we also need to add WPT tests for this update, do you have anything inflight?

File: index.html [hunk 1]
```diff
@@ -416,6 +417,15 @@ <h2><dfn>getEntriesByName()</dfn> method</h2>
           </section>
         </section>
       </section>
+      <section>
+        <h2><dfn>takeRecords()</dfn> method</h2>
+        <p>The <a>takeRecords()</a> method instructs the user agent to flush  entries in <a>context object</a>'s <a>observer buffer</a> and must run these steps:</p>
+        <ol data-link-for="PerformanceObserver">
+          <li>Let <var>entries</var> be a copy of <a>context object</a>'s <a>observer buffer</a>.</li>
+          <li>Empty <a>context object</a>'s <a>observer buffer</a>.</li>
+          <li>Return <var>entries</var>.</li>
```
Comment (tdresser) [inline index.html]: Don't we need to convert the list of PerformanceEntry objects to a PerformanceEntryList?

File: index.html [hunk 2]
```diff
@@ -416,6 +417,15 @@ <h2><dfn>getEntriesByName()</dfn> method</h2>
           </section>
         </section>
       </section>
+      <section>
+        <h2><dfn>takeRecords()</dfn> method</h2>
+        <p>The <a>takeRecords()</a> method instructs the user agent to flush  entries in <a>context object</a>'s <a>observer buffer</a> and must run these steps:</p>
```
Comment (tdresser) [inline index.html]: flush  entries -> flush entries

in <a>context object -> in the <a>context object

I'm a bit confused by this.
Is this equivalent to saying 
"and must run these steps" -> "by running these steps"?

Comment (igrigorik) [reply index.html]: Ack. 

> "and must run these steps" -> "by running these steps"?

The preamble is short explainer for the method, and the algorithm are the actual instructions that the UA must execute; we use this pattern in other other places in the spec.

Comment (npm1) [reply index.html]: I agree that "by running these steps" is clearer. We could modify those other places in a followup.

**Comment by npm1** [comment]
We do not currently have takeRecords tests, so I'll need to add some. Shouldn't be hard to do.

File: index.html (hunk 1 above)
Comment (igrigorik) [reply index.html]: Those are equivalent "typedef sequence<PerformanceEntry> PerformanceEntryList;"

That said, agree that this should be clearer, will update the definition.

**Comment by igrigorik** [comment]
@tdresser @npm1 updated and simplified to align with other definitions.

File: index.html (hunk 1 above)
Comment (tdresser) [reply index.html]: Oops, that makes sense, thanks!

File: index.html [hunk 3]
```diff
@@ -416,6 +417,12 @@ <h2><dfn>getEntriesByName()</dfn> method</h2>
           </section>
         </section>
       </section>
+      <section>
+        <h2><dfn>takeRecords()</dfn> method</h2>
+        <p>The <a>takeRecords()</a> method must return a copy of the
+          <a>context object</a>'s <a>observer buffer</a>, and also empty
```
Comment (tdresser) [inline index.html]: Let me know if this is being overly picky.

I'd prefer "and then empty" vs "and also empty", as "also" doesn't as strongly imply an order to these operations.

File: index.html [hunk 4]
```diff
@@ -416,6 +417,12 @@ <h2><dfn>getEntriesByName()</dfn> method</h2>
...
+          <a>context object</a>'s <a>observer buffer</a>.</p>
```
Comment (tdresser) [inline index.html]: the \<a>context ...

**Comment by npm1** [comment]
lgtm

**Comment by npm1** [comment]
I have a question that might be worth clarifying in the spec: what happens when takeRecords() is called during a callback of the observer?

**Comment by npm1** [comment]
Ping, is this ready to go?

**Comment by toddreifsteck** [comment]
@rniwa @wanderview @JosephPecoraro  are any of you able to take a look and ensure there is nothing you are fundamentally against here?

I have no pushback on the concept.

**Comment by rniwa** [comment]
Seems fine assuming this works like `MutationObserver`'s `takeRecords`.

**Comment by igrigorik** [comment]
> I have a question that might be worth clarifying in the spec: what happens when takeRecords() is called during a callback of the observer?

Practically speaking, I think this will be a no-op in most cases (as in, it'll return an empty entrilyst) as long as the callback is not taking its sweet time to complete: the entrylist passed to the callback should already have the latest copy of everything in the observer's buffer. That said, I guess it is possible that in the time delta between when observer's callback receives its list and when takeRecords is called, some new entry is queued.. in that case, the observer's buffer is flushed synchronously and will return what's in the buffer. 

As far as processing logic, I don't think there is anything special here? Let me know if otherwise.

> @rniwa: Seems fine assuming this works like MutationObserver's takeRecords.

Yep, that's the intent.

---

@toddreifsteck @rniwa @tdresser thanks for the reviews. I think this is good to merge as soon as we address the last question by @npm1.

**Comment by npm1** [comment]
I don't recall if I had any special case in mind. Looking at the descriptions, it's clear that takeRecords will return nil (or any entries observed after starting the callback). So still LGTM.

**Comment by igrigorik** [comment]
Great, then I think we're good to merge. Thanks everyone!

**Pull request merged**